File: backend/.backups/full-snapshot-20260224-latest/scripts/download-headshots.js
#!/usr/bin/env node
/**
 * Download headshots for quote authors from Wikimedia Commons and other sources.
 * Saves to public/headshots/ for use in the app.
 */

const https = require("https");
const http = require("http");
const fs = require("fs");
const path = require("path");

const OUTPUT_DIR = path.join(__dirname, "../public/headshots");

// All headshot URLs - Wikimedia Commons (free to use) + any other verified sources
const HEADSHOTS = {
  "Warren Buffett":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/Warren_Buffett_KU_Visit.jpg/440px-Warren_Buffett_KU_Visit.jpg",
  "Charlie Munger":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/5/56/Charlie_Munger_%28cropped%29.jpg/440px-Charlie_Munger_%28cropped%29.jpg",
  "Benjamin Graham":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/2/2a/Benjamin_Graham_%281894-1976%29_portrait_on_23_March_1950.jpg/440px-Benjamin_Graham_%281894-1976%29_portrait_on_23_March_1950.jpg",
  "John Maynard Keynes":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/6/66/John_Maynard_Keynes.jpg/440px-John_Maynard_Keynes.jpg",
  "George Soros":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/5/5f/George_Soros_-_Festival_Economia_2012.JPG/440px-George_Soros_-_Festival_Economia_2012.JPG",
  "Ray Dalio":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/c/c1/Ray_Dalio_2017_%28cropped%29.jpg/440px-Ray_Dalio_2017_%28cropped%29.jpg",
  "Peter Lynch":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/e/e4/Peter_Lynch_%28cropped%29.jpg/440px-Peter_Lynch_%28cropped%29.jpg",
  "Howard Marks":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Howard_Marks_%28investor%29.jpg/440px-Howard_Marks_%28investor%29.jpg",
  "Paul Tudor Jones":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a1/Paul_Tudor_Jones_at_the_Robin_Hood_Foundation.jpg/440px-Paul_Tudor_Jones_at_the_Robin_Hood_Foundation.jpg",
  "Mellody Hobson":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/7/7d/Mellody_Hobson.jpg/440px-Mellody_Hobson.jpg",
  "Edwin Lefevre":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Edwin_Lef%C3%A8vre.jpg/440px-Edwin_Lef%C3%A8vre.jpg",
  "Robert Arnott":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/2/2d/Rob_Arnott.jpg/440px-Rob_Arnott.jpg",
  "Nicolas Darvas":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b0/Nicolas_Darvas.jpg/440px-Nicolas_Darvas.jpg",
  "Jesse Livermore":
    "https://upload.wikimedia.org/wikipedia/commons/thumb/0/0a/Press_photo_of_Jesse_Livermore_in_1940_%28cropped%29.jpg/440px-Press_photo_of_Jesse_Livermore_in_1940_%28cropped%29.jpg",
};

function slug(name) {
  return name
    .replace(/[.'']/g, "")
    .replace(/\s+/g, "-")
    .toLowerCase();
}

function download(url) {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith("https") ? https : http;
    protocol
      .get(
        url,
        {
          headers: {
            "User-Agent":
              "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36",
          },
        },
        (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
          return download(res.headers.location).then(resolve).catch(reject);
        }
        if (res.statusCode !== 200) {
          reject(new Error(`HTTP ${res.statusCode}`));
          return;
        }
        const chunks = [];
        res.on("data", (c) => chunks.push(c));
        res.on("end", () => resolve(Buffer.concat(chunks)));
      })
      .on("error", reject);
  });
}

async function main() {
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  const results = [];
  const delay = (ms) => new Promise((r) => setTimeout(r, ms));
  for (const [name, url] of Object.entries(HEADSHOTS)) {
    const filename = `${slug(name)}.jpg`;
    const filepath = path.join(OUTPUT_DIR, filename);
    try {
      const data = await download(url);
      fs.writeFileSync(filepath, data);
      results.push({ name, status: "ok", file: filename });
      console.log(`✓ ${name}`);
    } catch (err) {
      results.push({ name, status: "fail", error: err.message });
      console.log(`✗ ${name}: ${err.message}`);
    }
    await delay(1500); // Avoid rate limiting
  }

  const ok = results.filter((r) => r.status === "ok").length;
  const fail = results.filter((r) => r.status === "fail").length;
  console.log(`\nDone: ${ok} downloaded, ${fail} failed.`);
  console.log(`Saved to: ${OUTPUT_DIR}`);
}

main().catch(console.error);

File: scripts/download_headshots.py
#!/usr/bin/env python3
"""
Download headshots for quote authors from DuckDuckGo image search.
Saves to public/headshots/ for use in the app.
Run: python scripts/download_headshots.py
"""

import os
import re
import time
import urllib.request
from pathlib import Path

# All quote authors who need headshots
AUTHORS = [
    "Warren Buffett",
    "Charlie Munger",
    "Benjamin Graham",
    "John Maynard Keynes",
    "George Soros",
    "Ray Dalio",
    "Peter Lynch",
    "Howard Marks",
    "Paul Tudor Jones",
    "Mellody Hobson",
    "Edwin Lefevre",
    "Sir John Templeton",
    "Ed Seykota",
    "Alexander Elder",
    "William J. O'Neil",
    "Philip Fisher",
    "Robert Arnott",
    "Victor Sperandeo",
    "Robert Olstein",
    "Ken Fisher",
    "Larry Hite",
    "Nicolas Darvas",
    "Jesse Livermore",
    "Morgan Housel",
]

OUTPUT_DIR = Path(__file__).parent.parent / "public" / "headshots"


def slug(name: str) -> str:
    return re.sub(r"[.'\s]+", "-", name).strip("-").lower()


def search_and_download(name: str, ddgs=None) -> bool:
    """Use DuckDuckGo to find and download a headshot."""
    try:
        from duckduckgo_search import DDGS

        if ddgs is None:
            ddgs = DDGS()
        results = list(ddgs.images(f"{name} headshot portrait", max_results=10))
    except ImportError:
        print("  Install: pip install duckduckgo-search")
        return False

    if not results:
        print(f"  No results for {name}")
        return False

    for r in results:
        url = r.get("image")
        if not url or not url.startswith("http"):
            continue
        try:
            req = urllib.request.Request(
                url,
                headers={
                    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)"
                },
            )
            with urllib.request.urlopen(req, timeout=15) as resp:
                data = resp.read()
            if len(data) < 5000:
                continue
            out = OUTPUT_DIR / f"{slug(name)}.jpg"
            out.write_bytes(data)
            print(f"  ✓ {name}")
            return True
        except Exception as e:
            continue
    print(f"  ✗ {name} (failed)")
    return False


def main():
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    print(f"Saving to {OUTPUT_DIR}\n")
    ok = 0
    try:
        try:
            from ddgs import DDGS
        except ImportError:
            from duckduckgo_search import DDGS

        ddgs = DDGS()
    except ImportError:
        ddgs = None
    for i, name in enumerate(AUTHORS):
        out_jpg = OUTPUT_DIR / f"{slug(name)}.jpg"
        out_png = OUTPUT_DIR / f"{slug(name)}.png"
        if out_jpg.exists() and out_jpg.stat().st_size > 10000:
            print(f"[{i+1}/{len(AUTHORS)}] {name} (skip, exists)")
            ok += 1
            continue
        if out_png.exists() and out_png.stat().st_size > 10000:
            print(f"[{i+1}/{len(AUTHORS)}] {name} (skip, exists)")
            ok += 1
            continue
        print(f"[{i+1}/{len(AUTHORS)}] {name}")
        for attempt in range(2):
            try:
                if search_and_download(name, ddgs):
                    ok += 1
                break
            except Exception as e:
                if "Ratelimit" in str(e) and attempt == 0:
                    print(f"  Rate limited. Waiting 60s...")
                    time.sleep(60)
                else:
                    print(f"  Error: {e}")
                    break
        time.sleep(5)
    print(f"\nDone: {ok}/{len(AUTHORS)} downloaded")


if __name__ == "__main__":
    main()
